README.md
[![Build Status](https://travis-ci.org/acsone/pycoda.svg?branch=master)](https://travis-ci.org/acsone/pycoda)
[![Coverage Status](https://coveralls.io/repos/acsone/pycoda/badge.png?branch=master)](https://coveralls.io/r/acsone/pycoda?branch=master)

pycoda
======

A python library for parsing CODA files


The parsing algorythm is based on a code originally part of an openerp module wittren by Noviat nv/sa (www.noviat.be). 

Authors:

 * Laurent Mignon (ACSONE)


Please don't hesitate to give feedback and especially report bugs or
ask for new features through launchpad at this URL: https://bugs.launchpad.net/pycoda/+bugs



CHANGES.rst
Changes
~~~~~~~

1.1.0 (2022-05-18)
------------------
- `github PR11 <https://github.com/acsone/pycoda/pull/11>`_:
  Store the move reference and the move reference detail into separate fields
  to ease reconciliation.


1.0.0 (2018-09-26)
------------------

- Compatible python 3 (tested with 3.5)


0.5 (2018-02-06)
----------------

- Add support for foreign account number.


0.4 (2015-02-26)
----------------

- `github ISSUE3 <https://github.com/acsone/pycoda/issues/3>`_:
  Globalisation lines are considered as Normal if they are not
  followed by details lines. The transaction type is now used
  to detect the details lines.


0.3 (2015-01-15)
----------------

- `github PR2 <https://github.com/acsone/pycoda/pull/2>`_:
  fix error in parseInformationRecord (@vparitskiy)


0.2 (2014-09-15)
----------------

- `github PR1 <https://github.com/acsone/pycoda/pull/1>`_:
  bugfix in exeption msg (@vparitskiy)

0.1 (2014-07-10)
----------------

- Initial implementation of the CODA reader


coda/tests/test_parser.py
# -*- coding: utf-8 -*-
#
# Authors: Laurent Mignon
# Copyright (c) 2013 Acsone SA/NV (http://www.acsone.eu)
# All Rights Reserved
#
# WARNING: This program as such is intended to be used by professional
# programmers who take the whole responsibility of assessing all potential
# consequences resulting from its eventual inadequacies and bugs.
# End users who are looking for a ready-to-use solution with commercial
# guarantees and support are strongly advised to contact a Free Software
# Service Company.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
#
#
from coda.parser import Parser, CodaParserException
from coda.statement import AmountSign, MovementRecordType
from nose.tools import eq_, assert_raises
import os

BASEPATH = os.path.dirname(__file__)


class TestParser(object):

    def _test_first_statement(self, statement):
        eq_(len(statement.movements), 32)
        eq_(len(statement.informations), 3)
        eq_(len(statement.free_comunications), 0)
        eq_(statement.creation_date, '2009-03-05')
        eq_(statement.separate_application, "00000")
        eq_(statement.version, "2")
        eq_(statement.acc_number, "BE86407051416150")
        eq_(statement.currency, "EUR")
        eq_(statement.description, "KBC-Bedrijfsrekening")
        eq_(statement.old_balance, 0.0)
        eq_(statement.old_balance_amount_sign, AmountSign.CREDIT)
        eq_(statement.old_balance_date, '2009-03-04')
        eq_(statement.account_holder_name, 'STORA ENSO LANGERBRUGGE NV')
        eq_(statement.paper_seq_number, "042")
        eq_(statement.coda_seq_number, "002")
        eq_(statement.new_balance, 0.0)
        eq_(statement.new_balance_amount_sign, AmountSign.CREDIT)
        eq_(statement.new_balance_date, '2009-03-05')
        eq_(statement.new_balance_paper_seq_number, "042")
        self._checkMovement(statement.movements[0])
        self._checkInformation(statement.informations[0])

    def test_single_statemet_parsing(self):
        parser = Parser()
        with open(os.path.join(BASEPATH,
                               "Coda_v2_3_single_statement.txt"), 'rb') as f:
            content = f.read()
        statements = parser.parse(content)
        assert len(statements) == 1
        # the first statement is the same for single and multi-statements file
        statement = statements[0]
        self._test_first_statement(statement)

    def test_multi_statemets_parsing(self):
        parser = Parser()
        with open(os.path.join(BASEPATH,
                               "Coda_v2_3_multi_statements.txt"), 'rb') as f:
            content = f.read()
        statements = parser.parse(content)
        assert len(statements) == 2
        self._test_first_statement(statements[0])

    def _checkMovement(self, movement):
        eq_(movement.ref_move_detail, "0000")
        eq_(movement.transaction_ref, "SWJVZ0BN6 BKTBBNPOSKZ")
        eq_(movement.transaction_amount, 20.0)
        eq_(movement.transaction_amount_sign,  AmountSign.CREDIT)
        eq_(movement.transaction_type, 0)
        eq_(movement.transaction_date, '2009-03-05')
        eq_(movement.transaction_family, "04")
        eq_(movement.transaction_code, "50")
        eq_(movement.transaction_category, "000")
        eq_(movement.communication_is_structured, True)
        eq_(movement.communication_type, "114")
        eq_(movement.communication, "+++931/3843/84900+++ 2905172259460041")
        eq_(movement.entry_date, '2009-03-05')
        eq_(movement.type, MovementRecordType.NORMAL)
        eq_(movement.globalisation_code, 0)
        eq_(movement.payment_reference, '')
        eq_(movement.counterparty_bic, '')
        eq_(movement.counterparty_number, None)
        eq_(movement.counterparty_name, None)
        eq_(movement.counterparty_address, None)
        eq_(movement.counterparty_currency, None)

    def _checkInformation(self, information):
        eq_(information.ref, '00100001')
        eq_(information.transaction_ref, 'SWJVZ0BOX DOMUCVDIU01')
        eq_(information.transaction_type, '0')
        eq_(information.transaction_family, "05")
        eq_(information.transaction_code, "01")
        eq_(information.transaction_category, "000")
        eq_(information.communication, "001AMERICAN EXPRESS")

    def test_wrong_globalisation(self):
        """Test wrong globalisation

        Check that a globalisation line without the corresponding
        'end globalisation' is considered as Normal
        """
        parser = Parser()
        with open(os.path.join(
                BASEPATH, "Coda_v2_3_faulty_globalisation.txt"), 'rb') as f:
            content = f.read()
        statements = parser.parse(content)
        eq_(len(statements), 1)
        statement = statements[0]
        eq_(len(statement.movements), 11)
        for mv in statement.movements:
            eq_(mv.type, MovementRecordType.NORMAL)

    def test_wrong_globalisation_2(self):
        """Test wrong globalisation

        Check that a globalisation line without details
        is considered as Normal (file contains only globalisation statemnts)
        """
        parser = Parser()
        with open(os.path.join(
                BASEPATH, "Coda_v2_3_faulty_globalisation_2.txt"), 'rb') as f:
            content = f.read()
        statements = parser.parse(content)
        eq_(len(statements), 1)
        statement = statements[0]
        eq_(len(statement.movements), 2)
        for mv in statement.movements:
            eq_(mv.type, MovementRecordType.NORMAL)

    def test_globalisation(self):
        """Test that a globalisation line with details is considerd as
        Globalisation and its details as Normal
        """
        parser = Parser()
        with open(os.path.join(
                BASEPATH, "Coda_v2_3_globalisation.txt"), 'rb') as f:
            content = f.read()
        statements = parser.parse(content)
        eq_(len(statements), 1)
        statement = statements[0]
        eq_(len(statement.movements), 5)
        for idx, mv in enumerate(statement.movements):
            if idx == 2:
                eq_(mv.type, MovementRecordType.GLOBALISATION)
            else:
                eq_(mv.type, MovementRecordType.NORMAL)
        with open(os.path.join(BASEPATH,
                               "Coda_v2_3_globalisation_2.txt"), 'rb') as f:
            content = f.read()
        statements = parser.parse(content)
        assert len(statements) == 1
        statement = statements[0]
        eq_(len(statement.movements), 4)
        for mv in statement.movements:
            eq_(mv.type, MovementRecordType.NORMAL)

    def test_unsupported_version(self):
        parser = Parser()
        with open(os.path.join(
                BASEPATH, "Coda_faulty_version.txt"), 'rb') as f:
            content = f.read()
        with assert_raises(CodaParserException) as cm:
            parser.parse(content)
        eq_(cm.exception.code, ' R001')
        eq_(cm.exception.msg,
            'CODA V5 statements are not supported, please contact your bank')

    def test_parse_methods(self):
        parser = Parser()
        # test a invalid file name
        with assert_raises(ValueError):
            parser.parse_file('invalid_file_name')
        # test parsing from a path to a file
        parser = Parser()
        file_name = os.path.join(BASEPATH, "Coda_v2_3_single_statement.txt")
        statements = parser.parse_file(file_name)
        eq_(len(statements), 1)

        # test parsing from a file-like object
        parser = Parser()
        with open(os.path.join(file_name), 'rb'):
            statements = parser.parse_file(file_name)
            eq_(len(statements), 1)

        # test parsing an invalid content
        with assert_raises(ValueError) as cm:
            parser.parse(b'invalid_coda_content')
        ex = cm.exception
        eq_(str(ex), 'The given value is not a valid coda content')

    def test_foreign_account(self):
        parser = Parser()
        file_name = os.path.join(BASEPATH, "Coda_foreign_account.txt")
        statements = parser.parse_file(file_name)
        eq_(len(statements), 1)
        st = statements[0]
        eq_(st.acc_number, 'FR1234567890240924002304825')
        eq_(st.currency, 'EUR')


coda/statement.py
# -*- coding: utf-8 -*-
#
# Authors: Laurent Mignon
# Copyright (c) 2013 Acsone SA/NV (http://www.acsone.eu)
# All Rights Reserved
#
# WARNING: This program as such is intended to be used by professional
# programmers who take the whole responsibility of assessing all potential
# consequences resulting from its eventual inadequacies and bugs.
# End users who are looking for a ready-to-use solution with commercial
# guarantees and support are strongly advised to contact a Free Software
# Service Company.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
#


class AmountSign(object):
    CREDIT = "0"
    DEBIT = "1"


class MovementRecordType(object):
    NORMAL = "0"
    GLOBALISATION = "1"


class MovementRecord(object):
    """A movement record
    """

    def __init__(self):
        self.ref = None

        # Continuous sequence number Starts at 0001 and is increased by 1 for
        # each movement record referring to
        # another movement on the daily statement
        # of account.
        self.ref_move = None

        # Detail number for
        # each movement record for the same
        # continuous sequence number.
        self.ref_move_detail = None
        self.transaction_ref = None
        self.transaction_amount = None
        self.transaction_amount_sign = None
        self.transaction_type = None
        self.transaction_date = None
        self.transaction_family = None
        self.transaction_code = None
        self.transaction_category = None
        self.communication_is_structured = None
        self.communication_type = None
        self.communication = None
        self.entry_date = None
        self.type = None
        self.globalisation_code = None
        self.payment_reference = None
        self.counterparty_bic = None
        self.counterparty_number = None
        self.counterparty_name = None
        self.counterparty_address = None
        self.counterparty_currency = None


class InformationRecord(object):
    """ Information record
    """

    def __init__(self):
        self.ref = None
        # Continuous sequence number: must be identical to
        # the continuous sequence number of the movement
        # record to which the information record refers.
        self.ref_move = None
        # Detail number 
        self.ref_move_detail = None
        self.transaction_ref = None
        self.transaction_type = None
        self.transaction_family = None
        self.transaction_code = None
        self.transaction_category = None
        self.communication = None


class FreeCommunication(object):
    """ Free communication
    """
    def __init__(self):
        self.ref = None
        self.communication = None


class Statement(object):
    """Statement of account
    """

    def __init__(self):
        self.creation_date = None
        self.separate_application = None
        self.version = None
        self.acc_number = None
        self.currency = None
        self.description = None
        self.old_balance = None
        self.old_balance_amount_sign = None
        self.old_balance_date = None
        self.account_holder_name = None
        self.paper_seq_number = None
        self.coda_seq_number = None
        self.new_balance = None
        self.new_balance_amount_sign = None
        self.new_balance_date = None
        self.new_balance_paper_seq_number = None
        self.movements = []
        self.informations = []
        self.free_comunications = []


coda/parser.py
# -*- coding: utf-8 -*-
#
# Authors: Laurent Mignon
# Copyright (c) 2013 Acsone SA/NV (http://www.acsone.eu)
# All Rights Reserved
#
# WARNING: This program as such is intended to be used by professional
# programmers who take the whole responsibility of assessing all potential
# consequences resulting from its eventual inadequacies and bugs.
# End users who are looking for a ready-to-use solution with commercial
# guarantees and support are strongly advised to contact a Free Software
# Service Company.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
#
#
import os
import re
import time

from .statement import MovementRecord, MovementRecordType, InformationRecord, \
    FreeCommunication, Statement


class CodaParserException(Exception):

    """Exception raised for errors in the input.

    Attributes:
        code -- exception code
        msg  -- explanation of the error
    """

    def __init__(self, code, msg):
        self.code = code
        self.msg = msg


class Parser(object):

    """CODA file parser mapping line to Python objects
    """

    def __init__(self, date_format='%Y-%m-%d'):
        self.date_format = date_format

    def is_valid_coda(self, value):
        """ Check if the given value is a valid coda content
        :param: value: data to parse
        :type param: str
        :returms: True id valid False otherwise
        """
        # Matches the first 24 characters of a CODA file, as defined by
        # the febelfin specifications
        return re.match(r'0{5}\d{9}05[ D] {7}', value) is not None

    def parse_file(self, fp):
        """ Parse the given file
         :param: fp: the path to the file to parse or a valid file-like object
         :returms: return a list of Statement objects found in the input file
         :rtype: list
        """
        if hasattr(fp, 'read'):
            return self.parse(fp.read())
        elif os.path.exists(fp):
            with open(fp, 'rb') as f:
                return self.parse(f.read())
        else:
            raise ValueError('The given argument is not a valid file-like '
                             'object nor a valid path to an existing file.')

    def parse(self, value):
        """Parse the given value.
        :param: value: data to parse
        :type param: bytes
        :returms: return a list of Statement objects found in value
         :rtype: list
        """
        value_unicode = value.decode('windows-1252', 'strict')
        if not self.is_valid_coda(value_unicode):
            raise ValueError('The given value is not a valid coda content')
        recordlist = value_unicode.split('\n')
        statements = []
        statement = None
        for line in recordlist:
            if not line:
                pass
            elif line[0] == '0':
                self.__fixes_globalisation_without_details(
                    statement)
                # Begin of a new Bank statement
                statement = Statement()
                self._parseHeader(line, statement)
                statements.append(statement)

            elif line[0] == '1':
                # Statement details
                self._parseHeaderDetails(line, statement)
            elif line[0] == '2':
                self._parseMovementRecord(line, statement)
            elif line[0] == '3':
                self._parseInformationRecord(line, statement)
            elif line[0] == '4':
                self.parseFreeCommunication(line, statement)
            elif line[0] == '8':
                # new balance record
                self._parseNewBalanceRecord(line, statement)
            elif line[0] == '9':
                # trailer record
                pass
        self.__fixes_globalisation_without_details(statement)
        return statements

    def __fixes_globalisation_without_details(self, statement):
        """ Change the movement type from globalisation to normal for the last
        movements
        """
        if statement and statement.movements:
            mv = statement.movements[-1]
            if mv.type == MovementRecordType.GLOBALISATION:
                mv.type = MovementRecordType.NORMAL

    def _parseHeader(self, line, statement):
        statement.version = version = line[127]
        if version not in ['1', '2']:
            raise CodaParserException(
                ' R001', 'CODA V%s statements are not supported, please '
                'contact your bank' % statement.version)
        statement.creation_date = time.strftime(
            self.date_format, time.strptime(rmspaces(line[5:11]), '%d%m%y'))
        statement.separate_application = rmspaces(line[83:88])

    def _parseHeaderDetails(self, line, statement):
        if statement.version == '1':
            statement.acc_number = rmspaces(line[5:17])
            statement.currency = rmspaces(line[18:21])
        elif statement.version == '2':
            if line[1] == '0':  # Belgian bank account BBAN structure
                statement.acc_number = rmspaces(line[5:17])
                statement.currency = rmspaces(line[18:21])
            elif line[1] == '1':  # foreign bank account BBAN structure
                raise CodaParserException(
                    ' R1001', 'Foreign bank accounts with BBAN structure are '
                    'not supported ')
            elif line[1] == '2':  # Belgian bank account IBAN structure
                statement.acc_number = rmspaces(line[5:21])
                statement.currency = rmspaces(line[39:42])
            elif line[1] == '3':  # foreign bank account IBAN structure
                _val = line[5:42]
                statement.acc_number = rmspaces(_val[:-3])
                statement.currency = rmspaces(_val[-3:])
            else:  # Something else, not supported
                raise CodaParserException(
                    ' R1003', 'Unsupported bank account structure ')

        statement.description = rmspaces(line[90:125])
        statement.old_balance = float(rmspaces(line[43:58])) / 1000
        statement.old_balance_amount_sign = line[42]
        statement.old_balance_date = time.strftime(
            self.date_format, time.strptime(rmspaces(line[58:64]), '%d%m%y'))
        statement.account_holder_name = rmspaces(line[64:90])
        statement.paper_seq_number = rmspaces(line[2:5])
        statement.coda_seq_number = rmspaces(line[125:128])

    def _parseMovementRecord(self, line, statement):
        if line[1] == '1':
            # New statement line
            record = MovementRecord()
            record.ref = rmspaces(line[2:10])
            record.ref_move = rmspaces(line[2:6])
            record.ref_move_detail = rmspaces(line[6:10])
            record.transaction_ref = rmspaces(line[10:31])
            record.transaction_amount_sign = line[31]  # 0 = Credit, 1 = Debit
            record.transaction_amount = float(rmspaces(line[32:47])) / 1000
            record.transaction_type = int(line[53])
            record.transaction_date = time.strftime(
                self.date_format, time.strptime(
                    rmspaces(line[47:53]), '%d%m%y'))
            record.transaction_family = rmspaces(line[54:56])
            record.transaction_code = rmspaces(line[56:58])
            record.transaction_category = rmspaces(line[58:61])
            record.communication_is_structured = line[61] == '1'
            if record.communication_is_structured:
                # Structured communication
                record.communication_type = line[62:65]
                record.communication = '+++' + \
                    line[65:68] + '/' + line[68:72] + '/' + line[72:77] + '+++'
            else:
                # Non-structured communication
                record.communication = rmspaces(line[62:115])
            record.entry_date = time.strftime(
                self.date_format, time.strptime(
                    rmspaces(line[115:121]), '%d%m%y'))
            record.type = MovementRecordType.NORMAL

            if record.transaction_type in [1, 2, 3]:
                # here the transaction type is a globalisation
                # 1 is for globalisation from the customer
                # 2 is for globalisation from the bank
                record.type = MovementRecordType.GLOBALISATION

            # a globalisation record can be followed by details lines
            # if it's not the case, the globalisation record is considered
            # as normal. To determine if a globalisation record is followed
            # by details, on a new line starting with 21, we check if the
            # previous line is a globalisation record. If the current line is
            # not a details line (transaction_type > 3) we reset the
            # record.type to Normal
            prev_mvmt = statement.movements and statement.movements[-1] or None
            if prev_mvmt and \
                record.transaction_type < 4 and \
                    prev_mvmt.type == MovementRecordType.GLOBALISATION:
                prev_mvmt.type = MovementRecordType.NORMAL
            record.globalisation_code = int(line[124])
            statement.movements.append(record)
        elif line[1] == '2':
            record = statement.movements[-1]
            if record.ref[0:4] != line[2:6]:
                raise CodaParserException(
                    'R2004', 'CODA parsing error on movement data record 2.2, '
                    'seq nr %s!' % line[2:10])
            record.communication = join_communications(
                record.communication, rmspaces(line[10:63]))
            record.payment_reference = rmspaces(line[63:98])
            record.counterparty_bic = rmspaces(line[98:109])
        elif line[1] == '3':
            record = statement.movements[-1]
            if record.ref[0:4] != line[2:6]:
                raise CodaParserException(
                    'R2005', 'CODA parsing error on movement data record 2.3, '
                    'eq nr %s!' % line[2:10])
            if statement.version == '1':
                record.counterparty_number = rmspaces(line[10:22])
                record.counterparty_name = rmspaces(line[47:73])
                record.counterparty_address = rmspaces(line[73:125])
                record.counterparty_currency = ''
            else:
                if line[22] == ' ':
                    record.counterparty_number = rmspaces(line[10:22])
                    record.counterparty_currency = rmspaces(line[23:26])
                else:
                    record.counterparty_number = rmspaces(line[10:44])
                    record.counterparty_currency = rmspaces(line[44:47])
                record.counterparty_name = rmspaces(line[47:82])
                record.communication = join_communications(
                    record.communication, rmspaces(line[82:125]))
        else:
            # movement data record 2.x (x != 1,2,3)
            raise CodaParserException(
                'R2006', '\nMovement data records of type 2.%s are not '
                'supported ' % line[1])

    def _parseInformationRecord(self, line, statement):
        if line[1] == '1':
            infoLine = InformationRecord()
            infoLine.ref = rmspaces(line[2:10])
            infoLine.ref_move = rmspaces(line[2:6])
            infoLine.ref_move_detail = rmspaces(line[6:10])
            infoLine.transaction_ref = rmspaces(line[10:31])
            infoLine.transaction_type = line[31]
            infoLine.transaction_family = rmspaces(line[32:34])
            infoLine.transaction_code = rmspaces(line[34:36])
            infoLine.transaction_category = rmspaces(line[36:39])
            infoLine.communication = rmspaces(line[40:113])
            statement.informations.append(infoLine)
        elif line[1] == '2':
            infoLine = statement.informations[-1]
            if infoLine.ref != rmspaces(line[2:10]):
                raise CodaParserException(
                    'R3004', 'CODA parsing error on information data '
                    'record 3.2, seq nr %s!' % line[2:10])
            infoLine.communication += rmspaces(line[10:100])
        elif line[1] == '3':
            infoLine = statement.informations[-1]
            if infoLine.ref != rmspaces(line[2:10]):
                raise CodaParserException(
                    'R3005', 'CODA parsing error on information data '
                    'record 3.3, seq nr %s!' % line[2:10])
            infoLine.communication += rmspaces(line[10:100])

    def _parseNewBalanceRecord(self, line, statement):
        statement.new_balance_amount_sign = line[41]
        statement.new_balance_paper_seq_number = rmspaces(line[1:4])
        statement.new_balance = float(rmspaces(line[42:57])) / 1000
        statement.new_balance_date = time.strftime(
            self.date_format, time.strptime(rmspaces(line[57:63]), '%d%m%y'))

    def parseFreeCommunication(self, line, statement):
        comm_line = FreeCommunication()
        comm_line.ref = rmspaces(line[2:10])
        comm_line.communication = rmspaces(line[32:112])
        statement.free_comunications.append(comm_line)


def join_communications(c1, c2):
    if not c1:
        return c2
    if not c2:
        return c1
    if not c2.startswith(" "):
        return " ".join([c1, c2])
    return c1 + c2


def rmspaces(s):
    return " ".join(s.split())
